test/main/modules.js

```diff
@@ -147,3 +147,141 @@ QUnit.module( "Deprecated setup/teardown", {
+				assert.ok( true, "first outer module afterEach called" );
+			},
+			beforeEach: function( assert ) {
+				assert.ok( true, "first outer beforeEach called" );
+			}
+		},
+		function() {
+			QUnit.module( "first inner", {
+					afterEach: function( assert ) {
+						assert.ok( true, "first inner module afterEach called" );
+					},
+					beforeEach: function( assert ) {
+						assert.ok( true, "first inner module beforeEach called" );
+					}
+				},
+				function() {
+					QUnit.test( "in module, before- and afterEach called in out-in-out " +
+						"order",
+						function( assert ) {
+							var module = assert.test.module;
+							assert.equal( module.name,
+								"nested modules > first outer > first inner" );
+							assert.expect( 5 );
+						});
+				});
+			QUnit.test( "test after nested module is processed", function( assert ) {
+				var module = assert.test.module;
+				assert.equal( module.name, "nested modules > first outer" );
+				assert.expect( 3 );
+			});
+			QUnit.module( "second inner" );
+			QUnit.test( "test after non-nesting module declared", function( assert ) {
+				var module = assert.test.module;
+				assert.equal( module.name, "nested modules > first outer > second inner" );
+				assert.expect( 3 );
+			});
+		});
+	QUnit.module( "second outer" );
+	QUnit.test( "test after all nesting modules processed and new module declared",
+		function( assert ) {
+			var module = assert.test.module;
+			assert.equal( module.name, "nested modules > second outer" );
+		});
+});
+
+QUnit.test( "modules with nested functions does not spread beyond", function( assert ) {
+	assert.equal( assert.test.module.name, "pre-nested modules" );
+});
+
+QUnit.module( "contained suite arguments", function( hooks ) {
+	QUnit.test( "hook functions", function( assert ) {
+		assert.strictEqual( typeof hooks.beforeEach, "function" );
+		assert.strictEqual( typeof hooks.afterEach, "function" );
+	} );
+
+	QUnit.module( "outer hooks", function( hooks ) {
+		var beforeEach = hooks.beforeEach;
+		var afterEach = hooks.afterEach;
+
+		beforeEach( function( assert ) {
+			assert.ok( true, "beforeEach called" );
+		} );
+
+		afterEach( function( assert ) {
+			assert.ok( true, "afterEach called" );
+		} );
+
+		QUnit.test( "call hooks", function( assert ) {
+			assert.expect( 2 );
+		} );
+
+		QUnit.module( "stacked inner hooks", function( hooks ) {
+			var beforeEach = hooks.beforeEach;
+			var afterEach = hooks.afterEach;
+
+			beforeEach( function( assert ) {
+				assert.ok( true, "nested beforeEach called" );
+			} );
+
+			afterEach( function( assert ) {
+				assert.ok( true, "nested afterEach called" );
+			} );
+
+			QUnit.test( "call hooks", function( assert ) {
+				assert.expect( 4 );
+			} );
+		} );
+	} );
+} );
+
+QUnit.module( "contained suite `this`", function( hooks ) {
+	this.outer = 1;
+
+	hooks.beforeEach( function() {
+		this.outer++;
+	} );
+
+	hooks.afterEach( function( assert ) {
+		assert.equal( this.outer, 2 );
```

Comment:
this commit will be merged with 53778e3.

It is passing here. I just wanted to set the steps on different commits during the review stage.

> It is unfortunate, but IIRC tests within a module do currently share environment.

I've done it this way getting from this assertion below from your example.

```js
assert.ok(this.outer, "outer module initialized environment");
```

Should it be different?